images/daemon/pkg/daemon/kata_check_node.go

```diff
@@ -0,0 +1,180 @@
+
+package daemon
+
+import (
+    "io/ioutil"
+    "strings"
+    "regexp"
+    "fmt"
+    "path/filepath"
+    "os"
+    "os/exec"
+)
+
+const (
+	procCPUInfo  = "/proc/cpuinfo"
+	sysModuleDir = "/sys/module"
+	moduleParamDir = "parameters"
+	modProbeCmd  = "modprobe"
+
+    cpuFlagsTag    = "flags"      // nolint: varcheck, unused, deadcode
+    cpuVendorField = "vendor_id"  // nolint: varcheck, unused, deadcode
+    cpuModelField  = "model name" // nolint: varcheck, unused, deadcode
+)
+
+var (
+    requiredCPUAttribs = []string {"GenuineIntel"}
+
+    requiredCPUFlags = []string {"vmx", "lm", "sse4_1"}
+
+    requiredKernelModules = map[string]map[string]string {
+        "kvm": {},
+        "kvm_intel": {
+            "unrestricted_guest": "Y",
+        },
+        "vhost": {},
+        "vhost_net": {},
+    }
+)
+
+
+func getFileContents(file string) (string, error) {
+    bytes, err := ioutil.ReadFile(file)
+    if err != nil {
+        return "", err
+    }
+
+    return string(bytes), nil
+}
+
+func findAnchoredString(haystack string, needle string) bool {
+    if haystack == "" || needle == "" {
+        return false
+    }
+    pattern := regexp.MustCompile(`\b` + needle + `\b`)
+    return pattern.MatchString(haystack)
+}
+
+func checkCPU(cpuinfo string, attribs []string) bool {
+    for _, attrib := range attribs {
+        found := findAnchoredString(cpuinfo, attrib)
+        if !found {
+            return false
+        }
+    }
+    return true
+}
+
+func haveKernelModule(moduleName string) bool {
+    path := filepath.Join(sysModuleDir, moduleName)
+
+    // check if module is loaded already
+    if _, err := os.Stat(path); err == nil {
+        return true
+    }
+
+    // try loading the module (as kata-daemon we're root by definition
+    // so permissions shouldn't be a concern)
+    cmd := exec.Command("modprobe", moduleName)
+    if output, err := cmd.CombinedOutput(); err != nil {
+        fmt.Printf("'modprobe %s' failed: %s", moduleName, output)
+        return false
+    }
```

Comment:
To me, this is quite an intrusive thing to be done as part of the check.
Consider we can load the module, how do we ensure the module will loaded next time we boot the node up?

Does the operator do something to ensure the module is up and will ever be while we have kata installed? If not, I'd just error out and say, "hey, sorry, no can do."

Reply:
@fidencio This is what upstream kata-check does.  I'm pointing this out just as a rationale of this PR's current form, I don't have a strong opinion of this and I'm open to doing it either way.